sw/device/lib/dif/dif_pinmux.c

```diff
@@ -0,0 +1,147 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_pinmux.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "pinmux_regs.h"  // Generated.
+#include "sw/device/lib/base/mmio.h"
+
+// TODO - autogenerate?
+#define PINMUX_REGISTER_WIDTH_IN_BYTES (PINMUX_PARAM_REG_WIDTH / 8)
+
+// TODO - autogenerate?
+#define PINMUX_INSEL_RESET_DEFAULT 0
+#define PINMUX_OUTSEL_RESET_DEFAULT 0x20820820
+
+/**
+ * Sets the peripheral to the reset defaults.
+ */
+void dif_pinmux_reset(const dif_pinmux_t *pinmux) {
+  // Reset all of the insel registers (multireg) to the reset value.
+  for (int i = 0; i < PINMUX_PERIPH_INSEL_MULTIREG_COUNT; ++i) {
+    ptrdiff_t register_offset =
+        PINMUX_PERIPH_INSEL0_REG_OFFSET + (i * PINMUX_REGISTER_WIDTH_IN_BYTES);
+    mmio_region_write32(pinmux->base_addr, register_offset,
+                        PINMUX_INSEL_RESET_DEFAULT);
+  }
+
+  // Reset all of the outsel registers (multireg) to the reset value.
+  for (int i = 0; i < PINMUX_MIO_OUTSEL_MULTIREG_COUNT; ++i) {
+    ptrdiff_t register_offset =
+        PINMUX_MIO_OUTSEL0_REG_OFFSET + (i * PINMUX_REGISTER_WIDTH_IN_BYTES);
+    mmio_region_write32(pinmux->base_addr, register_offset,
+                        PINMUX_OUTSEL_RESET_DEFAULT);
+  }
+}
+
+dif_pinmux_init_error_t dif_pinmux_init(mmio_region_t base_addr,
+                                        dif_pinmux_t *pinmux) {
+  if (pinmux == NULL) {
+    return kDifPinmuxInitErrorInvalidArgument;
+  }
+
+  // Check whether the peripheral can be configured.
+  bool write_enabled = mmio_region_get_bit32(base_addr, PINMUX_REGEN_REG_OFFSET,
+                                             PINMUX_REGEN_WEN);
+  if (!write_enabled) {
+    return kDifPinmuxInitErrorRegistersLocked;
+  }
+
+  pinmux->base_addr = base_addr;
+  pinmux->write_enabled = write_enabled;
+
+  // Set the peripheral to the reset defaults.
+  dif_pinmux_reset(pinmux);
+
+  return kDifPinmuxInitErrorNone;
+}
+
+dif_pinmux_registers_lock_error_t dif_pinmux_registers_lock(
+    dif_pinmux_t *pinmux) {
+  if (pinmux == NULL) {
+    return kDifPinmuxRegistersLockErrorInvalidArgument;
+  }
+
+  mmio_region_write32(pinmux->base_addr, PINMUX_REGEN_REG_OFFSET, 0);
+  pinmux->write_enabled = false;
+
+  return kDifPinmuxRegistersLockErrorNone;
+}
+
+dif_pinmux_insel_set_error_t dif_pinmux_insel_set(
+    const dif_pinmux_t *pinmux, dif_pinmux_peripheral_in_t peripheral_input,
+    dif_pinmux_insel_t mio_select) {
+  if (pinmux == NULL) {
+    return kDifPinmuxInselSetErrorInvalidArgument;
+  }
+
+  if (!pinmux->write_enabled) {
+    return kDifPinmuxInselSetErrorRegistersLocked;
+  }
+
+  if (mio_select < kDifPinmuxInselTieToZero ||
+      mio_select > kDifPinmuxInselPadLast) {
+    return kDifPinmuxInselSetErrorBadMioSelect;
+  }
```

Comment:
DIF-specific style guide says that you don't need to check enums if the function can operate on the entire range. Also, checking `mio_select` but not checking `peripheral_input` is inconsistent. Please also see my comments regarding the enums in the header file.

Reply:
Now they are uint32_t typedefs, so upper range should be checked.